features/field-based-acl.md

```diff
@@ -0,0 +1,190 @@
+Field-based ACL
+================
+
+# API Design Overview
+
+Field-based ACL is a Skygear feature defining the accessibility of the fields
+of a Skygear Record type.
+
+## Accessibility levels
+
+The field-based accessibility is divided into 5 levels:
+
+- `ReadWrite`: fields that available for read and write
+- `Queryable`: Fields that can be used to perform any query predicates and
+  will be returned after query.
+- `Discoverable`: Fields that can only be used to perform equality query
+  predicate and will be returned in query.
+- `ReadOnly`: Fields that cannot be used to perform any query predicates but
+  will be returned in query.
+- `NoAccess`: Fields that cannot be used to perform any query predicates and
+  will not be returned in query.
+
+## Access Control Entity (ACE)
+
+The field-based ACL is an array of field-based Access Control Entities (ACE).
+Each field-based ACE model will contains:
+
+- `RecordType`: the record type the entity belongs to, wildcards are allowed
+- `UserRole`: the user role the entity apply to
+- `Field`: the record field the entity apply to, wildcards are allowed
+- `AccessLevel`: the accessibility level the entity defines
+
+By defining field-based ACL, the permission of users performing actions on a
+record field is defined. A user would have permission to perform an action on a
+record field if any ACEs grant the permission of the role that the user belongs
+to, i.e. allow-based policy.
+
+## User Roles
+
+Since multiple rules with different user roles can be applied to a field of a
+record type, the resolve order of different user roles is important. The
+resolve order will be:
+
+1. **Public**: any users with correct API key
+1. **Owner**: the owner of the record instance
+1. **Specific User**: the user specified by user ID
+1. **Reference Users**: the users specified in a field of the records
+1. **Defined Roles**: the user-defined roles
+1. **Any Users**: any logged-in users
+
+## Base ACL
+
+Since the field-based ACL is allow-based, the following entities would be added
+serving as a base ACL.
+
+| Class | UserRole | Field |   Level   |
+|-------|----------|-------|-----------|
+| *     | AnyUser  | *     | ReadWrite |
```

Comment:
This refers to all fields in all records. So even new fields or new records are added by JIT, this rule should also be applied.

> Is this implicit or explicit?

I think it is open to implementor. For explicit declaration, Skygear server would throw an exception if no rules is found during accessibility checking. But it gives a flexibility to update the base ACL. For implicit inference, Skygear server would make the assumption in code level and not open to update.